File: leetcode/rotate-list.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        length = 0
        ptr = head
        while ptr:
            ptr = ptr.next
            length += 1

        if length == 0 or k % length == 0:
            return head

        k = k % length
        dummy = ListNode(0, head)

        slow_ptr = head
        fast_ptr = head

        while fast_ptr:
            # this is -1 because we need the previous node
            if k > -1:
                fast_ptr = fast_ptr.next
                k -= 1
            else:
                slow_ptr = slow_ptr.next
                fast_ptr = fast_ptr.next

        dummy.next = slow_ptr.next
        slow_ptr.next = None
        ptr = dummy.next

        while ptr.next:
            ptr = ptr.next

        ptr.next = head

        return dummy.next

File: leetcode/rotate-list.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
 public:
    ListNode *rotateRight(ListNode *head, int k)
    {
        ListNode *slowptr, *fastptr;
        ListNode dummy;

        int length = 0;

        dummy = ListNode(0, head);

        fastptr = head;
        for (; fastptr; fastptr = fastptr->next) ++length;

        if (head && k % length > 0) {
            k = k % length;

            fastptr = slowptr = &dummy;

            for (; fastptr; fastptr = fastptr->next) {
                if (k <= -1) {
                    slowptr = slowptr->next;
                }
                --k;
            }

            dummy.next = slowptr->next;
            slowptr->next = nullptr;
            slowptr = dummy.next;

            while (slowptr->next) {
                slowptr = slowptr->next;
            }

            slowptr->next = head;
        }

        return dummy.next;
    }
};
